src/lib/kimyi_plus.rs

```diff
@@ -0,0 +1,503 @@
+// of, display, perform, and distribute the Software and make, use, sell, offer for sale, import,
+// export, have made, and have sold the Software and the Larger Work(s), and to sublicense the
+// foregoing rights on either these or other terms.
+//
+// This license is subject to the following condition: The above copyright notice and either this
+// complete permission notice or at a minimum a reference to the UPL must be included in all copies
+// or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING
+// BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+use std::cmp::Ordering;
+use std::collections::HashSet;
+use std::convert::{TryFrom, TryInto};
+use std::fmt::Debug;
+
+use cactus::Cactus;
+use cfgrammar::Symbol;
+use lrtable::{Action, StIdx};
+use pathfinding::astar_bag;
+
+use kimyi::{apply_repairs, Dist, PathFNode, r3is, r3ir, r3d, r3s_n};
+use parser::{Node, Parser, ParseRepair};
+
+const PARSE_AT_LEAST: usize = 4; // N in Corchuelo et al.
+const PORTION_THRESHOLD: usize = 10; // N_t in Corchuelo et al.
+const TRY_PARSE_AT_MOST: usize = 250;
+
+pub(crate) fn recover<TokId: Clone + Copy + Debug + TryFrom<usize> + TryInto<usize> + PartialEq>
+                     (parser: &Parser<TokId>, in_la_idx: usize, in_pstack: &mut Vec<StIdx>,
+                      mut tstack: &mut Vec<Node<TokId>>)
+                  -> (usize, Vec<Vec<ParseRepair>>)
+{
+    // This function implements an algorithm whose core is based on that from "LR error repair
+    // using the A* algorithm" by Ik-Soon Kim and Kwangkeun Yi. However, we extend it in several
+    // ways.
+    //
+    // The basic idea behind this implementation is to use the transition rules from Fig 9 (along
+    // with the altered version of R3S presented on p12) as a mechanism for dynamically calculating
+    // the neighbours of the current node under investigation. Unlike KimYi, who
+    // non-deterministically return a single repair, this variant evaluates all minimal cost
+    // repairs.
+
+    let mut start_cactus_pstack = Cactus::new();
+    for st in in_pstack.drain(..) {
+        start_cactus_pstack = start_cactus_pstack.child(st);
+    }
+
+    let dist = Dist::new(parser.grm, parser.sgraph, |x| parser.ic(Symbol::Term(x)));
+    let start_node = PathFNode{pstack: start_cactus_pstack.clone(),
+                               la_idx: in_la_idx,
+                               t: 1,
+                               repairs: Cactus::new(),
+                               cf: 0,
+                               cg: 0};
+    let (astar_cnds, _) = astar_bag(
```

Comment:
This is the longest method call I've ever seen. Is this really Rust style?

Reply:
If you can work out how to split it up... I'm all ears ;) If it's any consolation, I split this up as far as I could into multiple submethods, but the call to `astar_bag` is complex because it takes 3 closures.